### PR TITLE
Correctly generate state dict in MultiTokenPredictionBlock 

### PR DESCRIPTION
**Pull request by asolergi-nv**
# What does this PR do ?

`MultiTokenPredictionBlock` has a `torch.nn.ModuleList` that contains all the layers, like `TransformerBlock` & `MambaStack`. The correct way to generate the sharded state dict is by calling the sharded_state_dict method layer by layer, since the ModuleList container does not have this method & will directly call the state_dict method of each layer. 

Before this fix we were generating the state_dict with some components coming from the `sharded_state_dict = super().sharded_state_dict(prefix, sharded_offsets, metadata)` call that called the state_dict method from the ModuleList and other parts from the per layer [sharded_state_dict method of each layer](https://github.com/asolergi-nv/Megatron-LM/blob/e7617e314da326af70adfb7f19b4ebe59ba1fd49/megatron/core/transformer/multi_token_prediction.py#L1420-L1424).

`TransformerBlock` state dict generation

https://github.com/NVIDIA/Megatron-LM/blob/027e0f3669838f8866b220a0a85551bdb21b6831/megatron/core/transformer/transformer_block.py#L926-L950

`MambaStack` state dict generation

https://github.com/NVIDIA/Megatron-LM/blob/027e0f3669838f8866b220a0a85551bdb21b6831/megatron/core/ssm/mamba_block.py#L448-L467

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

Feel free to message or comment the @mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [9f28b8e](https://github.com/NVIDIA/Megatron-LM/pull/3624/commits/9f28b8e769ba5afc2ffbd35bdee8adc96c4e3187)

**Review by ananthsub — APPROVED** [review]

**Comment by jaredcasper** [comment]
Is this going to cause backward compatibility issues with existing checkpoints?

**Comment by asolergi-nv** [comment]
> Is this going to cause backward compatibility issues with existing checkpoints?

The only difference I see in the state dict (both keys & contents) is that before this fix we were adding `self_attention.core_attention._extra_state` which its just a empty tensor (TransformerEngine convention of ALWAYS adding _extra_state entry despite its a empty tensor). Megatron-LM is [dropping the _extra_state](https://github.com/NVIDIA/Megatron-LM/blob/107f6ae34e75b686d27fd1f4706cbc7be4b6f5c4/megatron/core/extensions/transformer_engine.py#L1465) entry when using the `sharded_state_dict` method

**Comment by jaredcasper** [comment]
> The only difference I see in the state dict (both keys & contents) is that before this fix we were adding `self_attention.core_attention._extra_state` which its just a empty tensor (TransformerEngine convention of ALWAYS adding _extra_state entry despite its a empty tensor). Megatron-LM is [dropping the _extra_state](https://github.com/NVIDIA/Megatron-LM/blob/107f6ae34e75b686d27fd1f4706cbc7be4b6f5c4/megatron/core/extensions/transformer_engine.py#L1465) entry when using the `sharded_state_dict` method

Sorry, not sure the significance of that line that is dropped the `_extra_state` has. Do you just mean that before it would have an empty `_extra_state` and now it won't? If so that doesn't answer my question, does that empty `_extra_state` in old checkpoints cause problems with loading older checkpoints?

**Comment by asolergi-nv** [comment]
> does that empty _extra_state in old checkpoints cause problems with loading older checkpoints?

This will depend primarily on the `--dist-ckpt-strictness` setting used by the user. The default value ([`assume_ok_unexpected`](https://github.com/NVIDIA/Megatron-LM/blob/bfd160ba1cb4f6fb11c6cd0b2bdeb85c50fc7f73/megatron/training/config/training_config.py#L486)) won't cause any problems; it simply loads the keys present in the distributed checkpoint into the `sharded_state_dict` (and crashes if any key is missing from the distributed checkpoint). However, if the user uses stricter policies ([`raise_all`](https://github.com/NVIDIA/Megatron-LM/blob/bfd160ba1cb4f6fb11c6cd0b2bdeb85c50fc7f73/megatron/core/dist_checkpointing/validation.py#L76C18-L76C27)), then it will raise an issue because there will be a key in the distributed checkpoint that hasn't been loaded into the `sharded_state_dict` (The [`*_ALL`](https://github.com/NVIDIA/Megatron-LM/blob/bfd160ba1cb4f6fb11c6cd0b2bdeb85c50fc7f73/megatron/core/dist_checkpointing/validation.py#L47) category).

Given that this `_extra_state` entry is just an empty tensor and shouldn't raise any issues by default, I think it's safe to merge this change.

**Comment by asolergi-nv** [comment]
/ok to test [fdd7b21](https://github.com/NVIDIA/Megatron-LM/pull/3624/commits/fdd7b21f12a01ba31ba1c094e8adefa3d337c803)
